Do not asssume `Item.obj` in 'skipping' plugin

vidartf opened the pull request:
See #2231 for discussion.

Not entirely sure how to best wire up a regression test for this, but the following code should do the trick once collected:

```python
class CustomItem(pytest.Item):
    def setup(self):
        marker = pytest.mark.xfail(True)
        self.add_marker(marker)

    def runtest(self):
        assert False
```

nicoddemus commented:
Hey @vidartf thanks for following up with this! 👍 

Could you use the code you suggest along with the other files necessary (`conftest.py` file with custom collection hook, etc) which proves the patch work? If so, we can add an integration test in `test_skipping`.

vidartf commented:
@nicoddemus I added a test, and after an initial mistake, it seems now to pass. Is this sufficient?

Note: The mistaken test strangely passed on some versions of python. Is this a problem with CI? You will have to check CI logs to see it, as I rebased to keep git history clean.

nicoddemus commented:
Very nice, thank you!
